slurm/src/slurmcc/cost.py

```diff
@@ -0,0 +1,256 @@
+
+    def get_job_rec_file(self) -> str:
+        return os.path.join(self.cache, f"sacct-{self.start}-{self.end}.out")
+
+    def get_queue_records(self) -> str:
+
+        _queue_rec_file = self.get_queue_rec_file()
+        if self.use_cache(_queue_rec_file):
+            return _queue_rec_file
+
+        cmd = f"{self.squeue} --json"
+        with open(_queue_rec_file, 'w') as fp:
+            output = run_command(cmd, stdout=fp)
+            if output.returncode:
+                log.error("could not read slurm queue")
+        return _queue_rec_file
+
+    def process_queue(self) -> dict:
+        running_jobs = {}
+        queue_rec = self.get_queue_records()
+        with open(queue_rec, 'r') as fp:
+            data = json.load(fp)
+
+        for job in data['jobs']:
+            if job['job_state'] != 'RUNNING' and job['job_state'] != 'CONFIGURING':
+                continue
+            job_id = job['job_id']
+            if job['admin_comment']:
+                running_jobs[job_id] = job['admin_comment']
+        return running_jobs
+
+    def fetch_job_records(self) -> str:
+
+        _job_rec_file = self.get_job_rec_file()
+        if self.use_cache(_job_rec_file):
+            return _job_rec_file
+        cmd = self._construct_command()
+        with open(_job_rec_file, 'w') as fp:
+            output = run_command(cmd, stdout=fp)
+            if output.returncode:
+                log.error("Could not fetch slurm records")
+        return _job_rec_file
+
+    def parse_admincomment(self, comment: str):
+
+        return json.loads(comment)
+
+    def get_output_format(self, azcost: azurecost):
+
+        az_fmt = azcost.get_job_format()
+        return namedtuple('out_fmt_t', list(self.slurm_fmt_t._fields + az_fmt._fields + self.c_fmt_t._fields))
+
+    def process_jobs(self, azcost: azurecost, jobsfp, out_fmt_t):
+
+        _job_rec_file = self.fetch_job_records()
+        running = self.process_queue()
+        fp = open(_job_rec_file, newline='')
+        reader = csv.reader(fp, delimiter='|')
+        writer = csv.writer(jobsfp, delimiter=',')
+
+        for row in map(self.in_fmt_t._make, reader):
+            self.stats.jobs += 1
+            if row.state == 'RUNNING' and int(row.jobid) in running:
+                admincomment = running[int(row.jobid)]
+                self.stats.running_jobs += 1
+            else:
+                admincomment = row.admincomment
+            try:
+                comment_d = self.parse_admincomment(admincomment)[0]
+                sku_name = comment_d['vm_size']
+                cpupernode = comment_d['pcpu_count']
+                region = comment_d['location']
+                spot = comment_d['spot']
+            except json.JSONDecodeError as e:
+                log.debug(f"Cannot parse admincomment job={row.jobid} cluster={row.cluster}")
+                self.stats.admincomment_err += 1
+                self.stats.unprocessed += 1
+                continue
```

Comment:
You might want to add - what if it is valid json but it is just missing the fields you need?
except KeyError as ke:
...

Reply:
Fixed